lib/vrl/stdlib/src/unnest.rs

```diff
@@ -0,0 +1,196 @@
+                ),
+            },
+            Example {
+                title: "variable target",
+                source: indoc! {r#"
+                    foo = {"hostname": "localhost", "events": [{"message": "hello"}, {"message": "world"}]}
+                    foo = unnest!(foo.events)
+                "#},
+                result: Ok(
+                    r#"[{"hostname": "localhost", "events": {"message": "hello"}}, {"hostname": "localhost", "events": {"message": "world"}}]"#,
+                ),
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let path = arguments.required_query("path")?;
+
+        Ok(Box::new(UnnestFn { path }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct UnnestFn {
+    path: expression::Query,
+}
+
+impl UnnestFn {
+    #[cfg(test)]
+    fn new(path: &str) -> Self {
+        use std::str::FromStr;
+
+        Self {
+            path: expression::Query::new(
+                expression::Target::External,
+                FromStr::from_str(path).unwrap(),
+            ),
+        }
+    }
+}
+
+impl Expression for UnnestFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let path = self.path.path();
+
+        let value: Value;
+        let target: Box<&dyn Target> = match self.path.target() {
+            expression::Target::External => Box::new(ctx.target()) as Box<_>,
+            expression::Target::Internal(v) => {
+                let v = ctx.state().variable(v.ident()).unwrap_or(&Value::Null);
+                Box::new(v as &dyn Target) as Box<_>
+            }
+            expression::Target::Container(expr) => {
+                value = expr.resolve(ctx)?;
+                Box::new(&value as &dyn Target) as Box<&dyn Target>
+            }
+            expression::Target::FunctionCall(expr) => {
+                value = expr.resolve(ctx)?;
+                Box::new(&value as &dyn Target) as Box<&dyn Target>
+            }
+        };
+
+        let root = target.get(&LookupBuf::root())?.unwrap_or(Value::Null);
+
+        let values = root
+            .get_by_path(path)
+            .cloned()
+            .ok_or(value::Error::Expected {
+                got: Kind::Null,
+                expected: Kind::Array,
+            })?
+            .try_array()?;
+
+        let events = values
+            .into_iter()
+            .map(|value| {
+                let mut event = root.clone();
+                event.insert_by_path(path, value);
+                event
+            })
+            .collect::<Vec<_>>();
+
+        Ok(Value::Array(events))
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        match state.target_type_def() {
+            Some(root_type_def) => root_type_def
+                .clone()
+                .invert_array_at_path(&self.path.path())
+                .fallible(),
```

Comment:
It's not entirely clear to me why we need to go through the root of the target to set the final type def of the return value of this function (in other words: I'm not entirely sure why we need to expose `target_type_def()` for this use-case)?

Reply:
unnest takes the root object and merges it into each element of the provided array. Example:

```
$ . = { "noog" : 42, "nork" : [1,2,3] }
{ "noog": 42, "nork": [1, 2, 3] }

$ unnest!(.nork)
[{ "noog": 42, "nork": 1 }, { "noog": 42, "nork": 2 }, { "noog": 42, "nork": 3 }]
```

So to calculate the type def resulting from this function it needs to know the type def of the root fields to merge them into the resulting array.

Reply:
Ah yes, that makes sense. And this works for variables as well, correct?

```coffee
$ foo = { "noog" : 42, "nork" : [1,2,3] }
{ "noog": 42, "nork": [1, 2, 3] }

$ unnest!(foo.nork)
[{ "noog": 42, "nork": 1 }, { "noog": 42, "nork": 2 }, { "noog": 42, "nork": 3 }]
```

Reply:
I should add some tests to be sure.

Reply:
Good catch, no it doesn't.

Reply:
It does now.